lib/ex_doc/language/source.ex

```diff
@@ -0,0 +1,261 @@
+    ## Find all records in ast and collect any fields with type declarations
+    records =
+      filtermap_ast(abst_code, nil, fn
+        {:attribute, anno, :record, {name, fields}} ->
+          {name,
+           fields
+           |> Enum.flat_map(fn
+             {:typed_record_field, record_field, type} ->
+               [{:type, anno, :field_type, [elem(record_field, 2), type]}]
+
+             record_field when elem(record_field, 0) == :record_field ->
+               [{:type, anno, :field_type, [elem(record_field, 2), {:type, anno, :term, []}]}]
+
+             _ ->
+               []
+           end)}
+
+        _ ->
+          nil
+      end)
+      |> Map.new()
+
+    ## Expand records in all specs, callbacks, types and opaques
+    filtermap_ast(abst_code, nil, fn
+      {:attribute, anno, kind, {mfa, ast}} when kind in [:spec, :callback] ->
+        ast = Enum.map(ast, &expand_records(&1, records))
+        {:attribute, anno, kind, {mfa, ast}}
+
+      {:attribute, anno, type, {name, ast, args}} when type in [:opaque, :type] ->
+        {:attribute, anno, type, {name, expand_records(ast, records), args}}
+
+      otherwise ->
+        otherwise
+    end)
+  end
+
+  defp expand_records(types, records) when is_list(types) do
+    Enum.map(types, &expand_records(&1, records))
+  end
+
+  defp expand_records({:ann_type, anno, [name, type]}, records) do
+    {:ann_type, anno, [name, expand_records(type, records)]}
+  end
+
+  ## When we encounter a record, we fetch the type definitions in the record and
+  ## merge then with the type. If there are duplicates we take the one in the type
+  ## declaration
+  defp expand_records({:type, anno, :record, [{:atom, _, record} = name | args]}, records) do
+    args =
+      (args ++ Map.get(records, record, []))
+      |> Enum.uniq_by(fn {:type, _, :field_type, [{:atom, _, name} | _]} -> name end)
+
+    ## We delete the record from the map so that recursive
+    ## record definitions are not expanded.
+    records = Map.delete(records, record)
+
+    {:type, anno, :record, expand_records([name | args], records)}
+  end
+
+  defp expand_records({type, anno, what, args}, records) when type in [:type, :user_type] do
+    {:type, anno, what, expand_records(args, records)}
+  end
+
+  defp expand_records({:remote_type, anno, [m, t, args]}, records) do
+    {:remote_type, anno, [m, t, expand_records(args, records)]}
+  end
+
+  defp expand_records(otherwise, _records) do
+    otherwise
+  end
+
+  @doc """
+  Get the basedir of a module
+
+  The basedir is the cwd of the Elixir/Erlang compiler when compiling the module.
+  All `-file` attributes in the module is relative to this directory.
+  """
+  def get_basedir(abst_code, module) do
+    ## We look for the first -file attribute to see what the source file that
+    ## was compiled is called.
+    filename =
+      Enum.find_value(abst_code, fn
+        {:attribute, _anno, :file, {filename, _line}} ->
```

Comment:
Don't we have to guarantee this comes _before_ the module attribute? Otherwise we may find the filename of a function. Or do both abstract code guarantee there will be a `-file` at the top? I don't remember. :)

Reply:
Actually, if you can add a small note here, it would be very helpful. Does Elixir and Erlang include it at the top or not? Do they include in relative or absolute format or both? This way if we ever want to untangle both languages for whatever reason, we can at least have some idea of which branch is needed for each. :)

Reply:
If there is no `-file`, then the code falls back to using the `[:compile][:source]` attribute.

Reply:
I am not worried about when there is no `-file`, I am worried about file used to denote the location of a specific function, rather than the whole module.

You could do this:

```
-file {"foo.eex", 1}.
function_generated_by_template(...) -> ....
```

In this case, the `-file` applies only to that function, and this style is much more common in Elixir. I _think_ Erlang requires the `-file` for the whole module to be before the module attribute. If that's the intent here, we should stop looking for the file as soon as we find we see the `-module` attribute (but it has been ages since I looked into this).

Reply:
Do you have a full example in Elixir that I can look at?

Reply:
Go to `ExDoc.Markdown` and add this before the `to_ast` function:

```
  @doc """
  Converts the given markdown document to HTML AST.
  """
  @file "gotcha.ex"
  def to_ast(text, opts \\ []) when is_binary(text) do
```

The docs generated will be the same.

Now add `raise "oops"` to the first line of the `to_ast` function and run `mix build`, you will get this:

```
Generated ex_doc app
Generating docs...
** (RuntimeError) oops
    (ex_doc 0.31.1-dev) gotcha.ex:37: ExDoc.Markdown.to_ast/2
```

See how the file information was used in the stacktrace?

Reply:
And, fwiw, I don't think Elixir is adding the file information to the AST, rather somewhere in the Erlang compiler the `-file` attribute is used for that.

Reply:
The abstract code for this example is:

```
{:ok,
 {ExDoc.Markdown,
  [
    abstract_code: {:raw_abstract_v1,
     [
       {:attribute, 1, :file, {~c"lib/ex_doc/markdown.ex", 1}},
       {:attribute, 1, :module, ExDoc.Markdown},
       {:attribute, 1, :compile, [:no_auto_import]},
       {:attribute, 19, :callback,
        {{:to_ast, 2},
         [
           {:type, {19, 45}, :fun,
            [
              {:type, {19, 45}, :product,
               [
                 {:remote_type, {19, 26},
                  [{:atom, 0, String}, {:atom, 0, :t}, []]},
                 {:remote_type, {19, 39},
                  [{:atom, 0, Keyword}, {:atom, 0, :t}, []]}
               ]},
              {:type, {19, 48}, :term, []}
            ]}
         ]}},
       {:attribute, 24, :callback,
        {{:available?, 0},
         [
           {:type, {24, 26}, :fun,
            [{:type, {24, 26}, :product, []}, {:type, {24, 29}, :boolean, []}]}
         ]}},
       {:attribute, 1, :export,
        [
          __info__: 1,
          get_markdown_processor: 0,
          put_markdown_processor: 1,
          to_ast: 1,
          to_ast: 2
        ]},
       {:attribute, 1, :spec,
        {{:__info__, 1},
         [
           {:type, 1, :fun,
            [
              {:type, 1, :product,
               [
                 {:type, 1, :union,
                  [
                    {:atom, 1, :attributes},
                    {:atom, 1, :compile},
                    {:atom, 1, :functions},
                    {:atom, 1, :macros},
                    {:atom, 1, :md5},
                    {:atom, 1, :exports_md5},
                    {:atom, 1, :module},
                    {:atom, 1, :deprecated},
                    {:atom, 1, :struct}
                  ]}
               ]},
              {:type, 1, :any, []}
            ]}
         ]}},
       {:function, 0, :__info__, 1,
        [
          {:clause, 0, [{:atom, 0, :module}], [], [{:atom, 0, ExDoc.Markdown}]},
          {:clause, 0, [{:atom, 0, :functions}], [],
           [
             {:cons, 0,
              {:tuple, 0,
               [{:atom, 0, :get_markdown_processor}, {:integer, 0, 0}]},
              {:cons, 0,
               {:tuple, 0,
                [{:atom, 0, :put_markdown_processor}, {:integer, 0, 1}]},
               {:cons, 0, {:tuple, 0, [{:atom, 0, :to_ast}, {:integer, 0, 1}]},
                {:cons, 0,
                 {:tuple, 0, [{:atom, 0, :to_ast}, {:integer, 0, ...}]},
                 {nil, 0}}}}}
           ]},
          {:clause, 0, [{:atom, 0, :macros}], [], [nil: 0]},
          {:clause, 0, [{:atom, 0, :struct}], [], [{:atom, 0, nil}]},
          {:clause, 0, [{:atom, 0, :exports_md5}], [],
           [
             {:bin, 0,
              [
                {:bin_element, 0,
                 {:string, 0, [98, 13, 150, 29, 223, 104, 220, 12, 166, ...]},
                 :default, :default}
              ]}
           ]},
          {:clause, 0, [{:match, 0, {:var, 0, :Key}, {:atom, 0, :attributes}}],
           [],
           [
             {:call, 0,
              {:remote, 0, {:atom, 0, :erlang}, {:atom, 0, :get_module_info}},
              [{:atom, 0, ExDoc.Markdown}, {:var, 0, :Key}]}
           ]},
          {:clause, 0, [{:match, 0, {:var, 0, :Key}, {:atom, 0, :compile}}], [],
           [
             {:call, 0,
              {:remote, 0, {:atom, 0, :erlang}, {:atom, 0, :get_module_info}},
              [{:atom, 0, ExDoc.Markdown}, {:var, 0, :Key}]}
           ]},
          {:clause, 0, [{:match, 0, {:var, 0, :Key}, {:atom, 0, :md5}}], [],
           [
             {:call, 0,
              {:remote, 0, {:atom, 0, :erlang}, {:atom, 0, :get_module_info}},
              [{:atom, 0, ExDoc.Markdown}, {:var, 0, :Key}]}
           ]},
          {:clause, 0, [{:atom, 0, :deprecated}], [], [nil: 0]}
        ]},
       {:function, 68, :find_markdown_processor, 0,
        [
          {:clause, 68, [], [],
           [
             {:call, 69, {:remote, 69, {:atom, 69, Enum}, {:atom, 69, :find}},
              [
                {:cons, 69, {:atom, 69, ExDoc.Markdown.Earmark}, {nil, 69}},
                {:fun, 69,
                 {:clauses,
                  [
                    {:clause, 69, [{:var, 69, :_module@1}], [],
                     [{:case, 70, {:call, ...}, [...]}]}
                  ]}}
              ]}
           ]}
        ]},
       {:function, 45, :get_markdown_processor, 0,
        [
          {:clause, 45, [], [],
           [
             {:case, 46,
              {:call, 46,
               {:remote, 46, {:atom, 46, Application}, {:atom, 46, :fetch_env}},
               [{:atom, 46, :ex_doc}, {:atom, 46, :markdown_processor}]},
              [
                {:clause, 47,
                 [
                   {:tuple, 47,
                    [
                      {:atom, 47, :ok},
                      {:tuple, 47, [{:var, 47, :_processor@1}, {:var, 47, ...}]}
                    ]}
                 ], [],
                 [
                   {:tuple, 47,
                    [{:var, 48, :_processor@1}, {:var, 48, :_options@1}]}
                 ]},
                {:clause, 50, [{:atom, 50, :error}], [],
                 [
                   {:match, 51, {:var, 51, :_processor@2},
                    {:case, 51, {:call, 51, ...}, [{...}, ...]}},
                   {:call, 52, {:atom, 52, :put_markdown_processor},
                    [{:tuple, 52, [...]}]},
                   {:tuple, 0, [{:var, 53, :_processor@2}, {nil, 0}]}
                 ]}
              ]}
           ]}
        ]},
       {:function, 60, :put_markdown_processor, 1,
        [
          {:clause, 60, [{:var, 60, :_processor@1}],
           [
             [
               {:call, 60,
                {:remote, 60, {:atom, 60, :erlang}, {:atom, 60, :is_atom}},
                [{:var, 60, :_processor@1}]}
             ]
           ],
           [
             {:call, 61, {:atom, 61, :put_markdown_processor},
              [{:tuple, 61, [{:var, 61, :_processor@1}, {nil, 61}]}]}
           ]},
          {:clause, 64,
           [{:tuple, 64, [{:var, 64, :_processor@1}, {:var, 64, :_options@1}]}],
           [],
           [
             {:call, 65,
              {:remote, 65, {:atom, 65, Application}, {:atom, 65, :put_env}},
              [
                {:atom, 65, :ex_doc},
                {:atom, 65, :markdown_processor},
                {:tuple, 65,
                 [{:var, 65, :_processor@1}, {:var, 65, :_options@1}]}
              ]}
           ]}
        ]},
       {:function, 74, :raise_no_markdown_processor, 0,
        [
          {:clause, 74, [], [],
           [
             {:call, 75,
              {:remote, 75, {:atom, 75, :erlang}, {:atom, 75, :error}},
              [
                {:call, 75,
                 {:remote, 75, {:atom, 75, RuntimeError},
                  {:atom, 75, :exception}},
                 [
                   {:bin, 75,
                    [{:bin_element, 75, {:string, 75, ...}, :default, :default}]}
                 ]},
                {:atom, 75, :none},
                {:cons, 75,
                 {:tuple, 75, [{:atom, 75, :error_info}, {:map, 75, [{...}]}]},
                 {nil, 75}}
              ]}
           ]}
        ]},
       {:attribute, 36, :file, {~c"gotcha.ex", 123}},
       {:function, 36, :to_ast, 1,
        [
          {:clause, 36, [{:var, [generated: true, location: 0], :_@1}], [],
           [
             {:call, 36, {:atom, 36, :to_ast},
              [{:var, [generated: true, location: 0], :_@1}, {nil, 36}]}
           ]}
        ]},
       {:attribute, 36, :file, {~c"gotcha.ex", 123}},
       {:function, 36, :to_ast, 2,
        [
          {:clause, 36, [{:var, 36, :_text@1}, {:var, 36, :_opts@1}],
           [
             [
               {:call, 36,
                {:remote, 36, {:atom, 36, :erlang}, {:atom, 36, :is_binary}},
                [{:var, 36, :_text@1}]}
             ]
           ],
           [
             {:call, 37,
              {:remote, 37, {:atom, 37, :erlang}, {:atom, 37, :error}},
              [
                {:call, 37,
                 {:remote, 37, {:atom, 37, RuntimeError},
                  {:atom, 37, :exception}},
                 [{:bin, 37, [{:bin_element, ...}]}]},
                {:atom, 37, :none},
                {:cons, 37, {:tuple, 37, [{:atom, ...}, {...}]}, {nil, 37}}
              ]},
             {:match, 38,
              {:tuple, 38, [{:var, 38, :_processor@1}, {:var, 38, :_options@1}]},
              {:call, 38, {:atom, 38, :get_markdown_processor}, []}},
             {:call, 39,
              {:remote, 39, {:var, 39, :_processor@1}, {:atom, 39, :to_ast}},
              [
                {:var, 39, :_text@1},
                {:call, 39, {:remote, 39, {:atom, ...}, {...}},
                 [{:var, 39, ...}, {:var, ...}]}
              ]}
           ]}
        ]}
     ]}
  ]}}
```

So answer my earlier question, Elixir always guarantees the file at the top. :)

Reply:
So from what I can tell, Elixir places all functions and macros with `@file` attributes at the end of the AST, so that works as it should. I suppose you don't have anything similar to `-include` in elixir?

I will add a comment explaining the way parsing the ast should work.

Reply:
There is no `-include` yeah. I guess it makes sense for us to place `-file` at the end, because it is not tied to a function but rather it is processed in order?

Reply:
Anyway, we always place one at the top, so we are good to go from the Elixir side either way, not sure about Erlang. :)

Reply:
>  because it is not tied to a function but rather it is processed in order?

yes indeed.

> Anyway, we always place one at the top, so we are good to go from the Elixir side either way, not sure about Erlang. :)

We also do that.